# Opening and Exploring the Data

A mobile carrier operates a country-wide system of Base Transceiver Stations (BTS). The BTS system is the part of a GSM network that is responsible for the reception and transmission of radio signals from mobile phones.

There were approxiamately 700 Base Transceiver  Station in Bandung. I found [a data set](http://data.bandung.go.id/dataset/data-menara-telekomunikasi-di-kota-bandung) containing data about approximately 700 BTS from [OpenData Kota Bandung](http://data.bandung.go.id/)

Let's start by opening the data set then continue with exploring the data.

In [1]:
from csv import reader

opened_file = open ('DataBTSBandung2017.csv')
read_file = reader (opened_file)
BTS = list (read_file)
BTS_Header = BTS [0]
BTS = BTS [1:]

To make it easier to explore data set, we'll first write a function named **explore_data()** that we can use repeatedly to explore rows in a more readable way. We'll also add an option for our function to show the number of rows and columns for any data set.

In [2]:
def explore_data (dataset, start, end, row_and_columns = False):
    dataset_slice = dataset [start:end]
    for row in dataset_slice:
        print (row)
        print ('\n')
        
    if row_and_columns:
        print ('Number of Rows: ', len(dataset))
        print ('Number of Columns: ', len(dataset[0]))

In [3]:
print (BTS_Header)
print ('\n')
explore_data (BTS, 0, 3, True)

['No.', 'Kecamatan', 'Kelurahan', 'ID Survey', 'Pemilik Menara', 'Lokasi', 'Operator', 'Longdesimal', 'Latdesimal', 'Tower Height', 'Building Heights', 'Struktur Tower', 'Kategori', 'Tahun Pendirian', 'Kelistrikan', 'Status Lahan', 'Keterangan Lain-lain']


['1', 'Gedebage', 'Rancabolang', 'Team 1 site 2', 'Tower Bersama Group', 'Jl. Gedebage Selatan', 'Telkomsel- Smartfren', '107.687.346', '-6.952.163', '72', '0', '4 Kaki', 'BTS', '2003', 'PLN + GENSET', 'Sewa', '-']


['2', 'Gedebage', 'Rancabolang', 'Team 1 site 3', 'Natrindo Telepon Seluler', 'Jl. Gedebage Rancabolang', 'Tidak Teridentifikasi', '107.686.927', '-695.375', '42', '0', '3 Kaki', 'BTS', '2008', 'PLN', 'Sewa', 'Sudah Tidak Aktif']


['3', 'Gedebage', 'Rancanumpang', 'Team 1 site 6', 'Tower Bersama Group', 'Jl. Rancanumpang Rt 05/07', 'Telkomsel', '107.704.759', '-6.964.942', '65', '0', '4 Kaki', 'BTS', '-', 'PLN', 'Sewa', '-']


Number of Rows:  667
Number of Columns:  17


We see that the BTS data set has 667 point and 17 columns. At a quick glance, the columns that might be useful for the purpose of our analysis are 'Kecamatan', 'Kelurahan', 'Operator', and 'Keterangan Lain-lain'.

# Filtering Data Entries

If we explore the data set long enough, we'll find that some BTS are no longer active. For instance, If we explore in column **Keterangan dan lain-lain** , then it will be found BTS no longer active:


In [4]:
for itr in BTS:
    status = itr[-1]
    if status == "Sudah Tidak Aktif":
        print (itr)
        print ('\n')

['2', 'Gedebage', 'Rancabolang', 'Team 1 site 3', 'Natrindo Telepon Seluler', 'Jl. Gedebage Rancabolang', 'Tidak Teridentifikasi', '107.686.927', '-695.375', '42', '0', '3 Kaki', 'BTS', '2008', 'PLN', 'Sewa', 'Sudah Tidak Aktif']


['7', 'Panyileukan', 'Mekarmulya', 'Team 1 site 15', 'Protelindo', 'Jl. Mekarmulya', 'Tidak Teridentifikasi', '107.699.515', '-6.935.509', '42', '0', 'Monopole', 'BTS', '-', 'PLN', 'Sewa', 'Sudah Tidak Aktif']


['9', 'Panyileukan', 'Cipadung Kulon', 'Team 1 Site 20', 'Natrindo Telepon Seluler', 'Jl. H. Nasution', 'Tidak Teridentifikasi', '107.709.813', '-6.924.835', '32', '0', '3 Kaki', 'BTS', '2008', 'PLN', 'Sewa', 'Sudah Tidak Aktif']


['17', 'Rancasari', 'Derwati', 'Team 1 Site 37', 'Natrindo Telepon Seluler', 'Jl. Derwati Rt 05/05', 'Tidak Teridentifikasi', '107.684.744', '-6.965.642', '32', '0', '3 Kaki', 'BTS', '-', 'PLN', 'Sewa', 'Sudah Tidak Aktif']


['20', 'Rancasari', 'Cipamokolan', 'Team 1 Site 44', 'Natrindo Telepon Seluler', 'Jl. Ranca Loa', 

As soon as we find the BTS does not have an operator:

In [5]:
for itr in BTS:
    status = itr[-1]
    if status == "Sudah Tidak Ada Operator":
        print (itr)
        print ('\n')

['335', 'Sumur Bandung', 'Kebon Pisang', 'Team 2 Site 219', 'Yogya Group', 'Jl. Sunda No. 83', 'Tidak teridentifikasi', '107.618.041', '-691.778', '25', '15', '4 Kaki', 'BTS', '-', 'PLN', 'Sewa', 'Sudah Tidak Ada Operator']


['345', 'Cicendo', 'Sukaraja', 'Team 2 Site 233', 'Lita Fm', 'Jl. Budi No. 42', 'Tidak teridentifikasi', '107.558.706', '-6.890.928', '45', '0', '4 Kaki', 'BTS', '-', 'PLN', 'Hak', 'Sudah Tidak Ada Operator']


['438', 'Bandung Wetan', 'Taman Sari', 'Team 2 Site 416', 'Tidak Teridentifikasi', 'Jl. Taman Sari', 'Tidak Teridentifikasi', '107.608.474', '-6.893.787', '26', '0', 'Microcell Pole', 'BTS', '-', 'PLN', 'Sewa', 'Sudah Tidak Ada Operator']


['496', 'Coblong', 'Lebak Gede', 'Team 2 Site 515', 'Pertamina', 'Jl. Raya Diponegoro', 'Tidak Teridentifikasi', '107.617.873', '-689.905', '25', '0', '4 Kaki', 'BTS', '-', 'PLN', 'Hak', 'Sudah Tidak Ada Operator']




Or we will find the BTS sealed:

In [6]:
for itr in BTS:
    status = itr[-1]
    if status == "Disegel":
        print (itr)
        print ('\n')

['466', 'Coblong', 'Lebak Siliwangi', 'Team 2 Site 454', 'Telkomsel', 'Jl. Siliwangi', 'Telkomsel', '107.612.824', '-6.885.215', '15', '0', 'Monopole', 'BTS', '-', 'PLN', 'Sewa', 'Disegel']


['566', 'Coblong', 'Dago', 'Team 2 sitetambahan_12', 'Telkomsel', 'Jl. Ir. H. Djuanda', 'Tidak Teridentifikasi', '107.619.178', '-6.874.449', '15', '-', 'Microcell Pole', 'BTS', '-', 'GENSET', 'Sewa', 'Disegel']




In [7]:
for itr in BTS:
    operator = itr[6]
    if operator == "-":
        print (itr)
        print ('\n')

['35', 'Buahbatu', 'Cijaura', 'Team 1 Site 72', 'Global Tower', 'Jl. Pasar Kordon', '-', '10.763.976', '-6.954.365', '32', '0', 'Kamuflase Minaret Di Masjid', 'BTS', '-', 'PLN', 'Sewa', 'Sudah Tidak Aktif']


['624', 'Coblong', 'Dago', 'Tambahan', 'Protelindo', 'Jl. Cisitu Indah Vl No.181 Rt.001/004', '-', '10.761.395', '-68.729', 'RT 12m', '-', 'Pole Rooftop', '-', '-', '-', '-', '-']


['625', 'Bojongloa Kidul', 'Mekar wangi', 'Tambahan', 'Protelindo', 'Jalan Tol Moh Toha', '-', '10.760.316', '-696.213', 'GF 27m', '-', 'Monopole', '-', '-', '-', '-', '-']


['626', 'Andir', 'kebon jeruk', 'Tambahan', 'Protelindo', ' Jl. Gardu Jati No. 81-83 Rt.02/07 . King Garden Hotel', '-', '10.759.788', '-69.174', 'RT 6m', '-', 'Pole Rooftop', '-', '-', '-', '-', '-']


['627', 'Cibeunying Kaler', 'Cigadung', 'Tambahan', 'Protelindo', 'Gg. Masjid Al-Ikhlas Rt. 02 Rw. 08', '-', '10.762.675', '-688.318', 'RT 6m', '-', 'Pole Rooftop', '-', '-', '-', '-', '-']


['628', 'Lengkong', 'CIJAGRA', 'Tambaha

We don't want to count BTS has no longer active when we analyze data, so we need to remove the inactive entries and keep only the active one. One thing we could do is remove the inactive rows manually and separate the list between active and inactive Base Transceiver Stations based on status above.

In [8]:
BTS_Active = []
BTS_Deactive = []

for itr in BTS:
    status = itr[-1]
    operator = itr[6]
    
    if (operator == "Tidak Teridentifikasi" or operator == "Tidak teridentifikasi") and (status == "Sudah Tidak Aktif"):
        BTS_Deactive.append(itr)
    elif (operator == "Tidak Teridentifikasi" or operator == "Tidak teridentifikasi") and (status == "Sudah Tidak Ada Operator"):
        BTS_Deactive.append(itr)
    elif operator == "-":
        BTS_Deactive.append(itr)
    else:
        BTS_Active.append(itr)

In [9]:
explore_data (BTS_Active, 0, 5, True)

['1', 'Gedebage', 'Rancabolang', 'Team 1 site 2', 'Tower Bersama Group', 'Jl. Gedebage Selatan', 'Telkomsel- Smartfren', '107.687.346', '-6.952.163', '72', '0', '4 Kaki', 'BTS', '2003', 'PLN + GENSET', 'Sewa', '-']


['3', 'Gedebage', 'Rancanumpang', 'Team 1 site 6', 'Tower Bersama Group', 'Jl. Rancanumpang Rt 05/07', 'Telkomsel', '107.704.759', '-6.964.942', '65', '0', '4 Kaki', 'BTS', '-', 'PLN', 'Sewa', '-']


['4', 'Panyileukan', 'Cipadung Kidul', 'Team 1 site 11', 'Xl Axiata', 'Jl. Cimencrang Rt 04/04', 'XL', '10.770.841', '-6.944.872', '32', '0', '3 Kaki', 'BTS', '-', 'PLN', 'Sewa', '-']


['5', 'Panyileukan', 'Cipadung Kidul', 'Team 1 site 13', 'Tower Bersama Group', 'Kp. Panyileukan Rt 03/01', 'Telkomsel- HCPT- Smartfren', '107.715.687', '-6.938.283', '42', '0', '4 Kaki', 'BTS', '2004', 'PLN', 'Sewa', '-']


['6', 'Panyileukan', 'Cipadung Kidul', 'Team 1 site 14', 'Xl Axiata', 'Jl. Soekarno Hatta No.754', 'XL', '107.710.083', '-6.936.429', '3', '16', 'Pole Rooftop', 'BTS', '-',

We see that the **BTS_Active** data set has 623 points and 17 columns.

In [10]:
explore_data (BTS_Deactive, 0, 5, True)

['2', 'Gedebage', 'Rancabolang', 'Team 1 site 3', 'Natrindo Telepon Seluler', 'Jl. Gedebage Rancabolang', 'Tidak Teridentifikasi', '107.686.927', '-695.375', '42', '0', '3 Kaki', 'BTS', '2008', 'PLN', 'Sewa', 'Sudah Tidak Aktif']


['7', 'Panyileukan', 'Mekarmulya', 'Team 1 site 15', 'Protelindo', 'Jl. Mekarmulya', 'Tidak Teridentifikasi', '107.699.515', '-6.935.509', '42', '0', 'Monopole', 'BTS', '-', 'PLN', 'Sewa', 'Sudah Tidak Aktif']


['9', 'Panyileukan', 'Cipadung Kulon', 'Team 1 Site 20', 'Natrindo Telepon Seluler', 'Jl. H. Nasution', 'Tidak Teridentifikasi', '107.709.813', '-6.924.835', '32', '0', '3 Kaki', 'BTS', '2008', 'PLN', 'Sewa', 'Sudah Tidak Aktif']


['17', 'Rancasari', 'Derwati', 'Team 1 Site 37', 'Natrindo Telepon Seluler', 'Jl. Derwati Rt 05/05', 'Tidak Teridentifikasi', '107.684.744', '-6.965.642', '32', '0', '3 Kaki', 'BTS', '-', 'PLN', 'Sewa', 'Sudah Tidak Aktif']


['30', 'Buahbatu', 'Cijaura', 'Team 1 Site 61', 'Bakrie Telekom', 'Jl. Cijaura Ilir No. 62', 'Tida

We see that the **BTS_Deactive** data set has 44 points and 17 columns.

# Number of BTS By District

We'll build two functions we can use to analyze the frequency tables:

* One function to generate frequency tables that show total
* Another function that we can use to display the number in a descending order


In [11]:
def freq_table (dataset, index):
    table = {}
    
    for row in dataset:
        value = row[index]
        if value in table:
            table[value] += 1
        else: 
            table[value] = 1
    return table

In [12]:
def display_table (dataset, index):
    table = freq_table (dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted (table_display, reverse = True)
    for entry in table_sorted:
        print (entry[1], ':', entry[0])

display_table (BTS_Active, 1)

Coblong : 56
Sumur Bandung : 37
Cicendo : 31
Regol : 30
Lengkong : 29
Cibenying Kidul : 28
Bandung Wetan : 27
Batu Nunggal : 25
Sukajadi : 24
Bandung Kulon : 24
Arcamanik : 24
Buahbatu : 23
Bandung Kidul : 23
Andir : 23
Babakan Ciparay : 22
Sukasari : 21
Cidadap : 21
Cibeunying Kaler : 18
Bojongloa Kidul : 18
Antapani : 17
Kiaracondong : 14
Cibiru : 14
Rancasari : 12
Mandalajati : 12
Astana Anyar : 11
Cinambo : 10
Ujungberung : 9
Panyileukan : 8
Bojongloa Kaler : 7
Gedebage : 5


We can see that there is a disparity in the number of BTS distributions in Bandung. Downtown areas such as Coblong (56), Bandung Well (37), Cicendo (31), and Regol (30) are still the largest contributions in the telecommunication network business in Bandung. Unlike Downtown Area, Suburbs has own problems, such as the lack of business, for suburbs such as Ujungberung (9), Panyileukan (8), Bojongloa Kaler (7), and Gedebage (5). 

# List of Operators in each sub-district

Now, we'll look at the distribution of operators in eac Sub District. For example we'll see a list of cellular operators in Buahbatu-District and Gedebage-Disctrict.

In [13]:
Headers = "Kelurahan -- Operator"
print(Headers)
for name in BTS_Active:
    if name[1] == 'Buahbatu':
        print (name[2], ':', name[6])

Kelurahan -- Operator
Margasari : Indosat
Cijaura : Telkomsel- HCPT
Margacinta : Telkomsel- HCPT
Cijaura : Telkomsel
Cijaura : Telkomsel
Cijaura : XL- Indosat
Margasari : XL
Cijaura : Smartfren
Margacinta : XL
Margasari : Telkomsel
Margasari : Telkomsel
Sekejati : HCPT- Smartfren- XL- Telkomsel
Sekejati : Telkomsel- Smartfren
Jatisari : Indosat
Jatisari : Telkomsel
Jatisari : Smartfren
Jatisari : Telkomsel
Sekajati : Telkomsel
Sekajati : XL
Sekejati : Telkomsel
Sekejati : XL- HCPT- Smartfren
Sekejati : Telkomsel
Cijagra : Telkomsel


In [14]:
print (Headers)
for name in BTS_Active:
    if name[1] == 'Gedebage':
        print (name[2], ':', name[6])

Kelurahan -- Operator
Rancabolang : Telkomsel- Smartfren
Rancanumpang : Telkomsel
Rancabolang : XL- Telkomsel- Smartfren- HCPT
Rancabolang : Telkomsel- Tidak Teridentifikasi
Cisaranten Kidul : Telkomsel


For the two examples above, we can conclude that Telkomsel still dominates wherever its territory.